### PR TITLE
refactor(webhook): change for mutating webhook to use Default to set ModelReg namespace

### PR DESCRIPTION
**Pull request by zdtsw**
## Description

- use Default() instead of setDSCDefaults()

## Merge criteria

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/webhook/webhook.go [hunk 1]
```diff
@@ -44,10 +47,10 @@ type OpenDataHubValidatingWebhook struct {
 	Decoder *admission.Decoder
 }
 
-func (w *OpenDataHubValidatingWebhook) SetupWithManager(mgr ctrl.Manager) {
+func (v *OpenDataHubValidatingWebhook) SetupValidatingWebhookWithManager(mgr ctrl.Manager) {
```
Comment (ykaliuta) [inline controllers/webhook/webhook.go:49]: Why?
They have distinct types and type.SetupWithManager is self describing

File: controllers/webhook/webhook.go [hunk 2]
```diff
@@ -76,7 +79,7 @@ func denyCountGtZero(ctx context.Context, cli client.Client, gvk schema.GroupVer
 	return admission.Allowed("")
 }
 
-func (w *OpenDataHubValidatingWebhook) checkDupCreation(ctx context.Context, req admission.Request) admission.Response {
+func (v *OpenDataHubValidatingWebhook) checkDupCreation(ctx context.Context, req admission.Request) admission.Response {
```
Comment (ykaliuta) [inline controllers/webhook/webhook.go:81]: why? w as webhook is good enough. (I would abstract it even further till s Self :)

File: controllers/webhook/webhook.go [hunk 3]
```diff
@@ -135,35 +138,43 @@ type OpenDataHubMutatingWebhook struct {
...
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	// convert to []byte
+	dscBytes, err := json.Marshal(dsc)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	// use PatchResponseFromRaw to ensure dsc CR is updated with defaults
+	return admission.PatchResponseFromRaw(req.Object.Raw, dscBytes)
```
Comment (ykaliuta) [inline controllers/webhook/webhook.go]: so finally this is the actual change :) But still, could you put some reasoning behind?

File: controllers/webhook/webhook.go [hunk 4]
```diff
@@ -128,42 +129,26 @@ func (w *OpenDataHubValidatingWebhook) Handle(ctx context.Context, req admission
...
-	resp = w.setDSCDefaults(ctx, dsc)
-	return resp
+func (m *DSCDefaulter) SetupMutateWebhookWithManager(mgr ctrl.Manager) {
+	mutateWebhook := admission.WithCustomDefaulter(mgr.GetScheme(), &dscv1.DataScienceCluster{}, m)
+	mgr.GetWebhookServer().Register("/mutate-opendatahub-io-v1", mutateWebhook)
 }
 
-func (w *OpenDataHubMutatingWebhook) setDSCDefaults(_ context.Context, dsc *dscv1.DataScienceCluster) admission.Response {
+// implement admission.CustomDefaulter interface.
+func (m *DSCDefaulter) Default(_ context.Context, obj runtime.Object) error {
```
Comment (bartoszmajsak) [inline controllers/webhook/webhook.go:144]: As this is now extracted to a simple function, how about providing some unit tests for it? It does not even have to employ EnvTest, plain unit testing should be fine here.

Comment (zdtsw) [reply controllers/webhook/webhook.go:145]: did some updates on the unit-test.
only for the case which mutate webhook is related.

**Review by ykaliuta — APPROVED** [review]

**Pull request merged**